pkg/db/jsonType.go

```diff
@@ -0,0 +1,60 @@
+package db
+
+import (
+	"database/sql"
+	"database/sql/driver"
+	"encoding/json"
+	"errors"
+
+	"github.com/justtrackio/gosoline/pkg/mdl"
+)
+
+var (
+	_                        driver.Valuer = JSONType[struct{}]{}
+	_                        sql.Scanner   = &JSONType[struct{}]{}
+	ErrJSONInvalidDriverType               = errors.New("invalid incoming driver type for json type")
+)
+
+type (
+	JSONType[T any] struct {
+		val        T
+		asJsonNull bool
+	}
+)
+
+func NewJSONType[T any](val T) JSONType[T] {
+	return JSONType[T]{
+		val: val,
+	}
+}
+
+func (t *JSONType[T]) AsJSONNull() JSONType[T] {
+	t.asJsonNull = true
+
+	return *t
+}
+
+func (t JSONType[T]) Get() T {
+	return t.val
+}
+
+func (t JSONType[T]) Value() (driver.Value, error) {
+	if mdl.IsNil(t.val) && !t.asJsonNull {
+		return nil, nil //nolint:nilnil // this is the expected behaviour by the driver package
+	}
+
+	return json.Marshal(t.val)
+}
+
+func (t *JSONType[T]) Scan(src any) error {
+	if src == nil {
+		return nil
+	}
+
+	switch src := src.(type) {
+	case []byte:
+		return json.Unmarshal(src, &t.val)
```

Comment:
Do I get this correct that we accept `NULL` from SQL as well as the string `null`, but the empty string is invalid? However, when I unmarshal the JSON `null` and my object is not a pointer, it will stay as it was initialized, right?

Reply:
For a SQL NULL we will produce the zero value for T, meaning if you want to preserve the semantic of that SQL NULL your T needs to be a pointer to your desired type.
The string null will also produce the zero value for T, however we only write this when configured on the writing part of JSONType.

The empty string as input to Value will produce `[]byte("\"\"")` as driver.Value. If you parse that again this is valid (even though afaik it is not valid json according to the json specs).

The empty string (as byte slice so `[]byte("")`) as input to Scan is indeed invalid since any empty file or file with only whitespace is not valid json.